### PR TITLE
Reload lambda code on version change

### PR DESCRIPTION
**Pull request by bhasin85**
Cloudformation deploy will reload the lambda code when the version parameter is changed.

dev stack is up in development account for testing. I am happy to show you the reload from my machine.

### CODE REVIEW

File: wps-cloudformation-template.yaml [hunk 1]
```diff
@@ -753,6 +757,104 @@ Resources:
...
+              - Effect: Allow
+                Action:
+                  - 'lambda:UpdateFunctionCode'
+                Resource: '*'
+  ReloadLambdaCodeFunction:
+    Type: "AWS::Lambda::Function"
+    Properties:
+      Code:
+        ZipFile: |
+          import os
```
Comment (lwgordonimos) [inline wps-cloudformation-template.yaml]: So many loops and prints ;) I pasted into an IDE and crunched it down a bit... untested, but a couple of list comprehensions removes some nesting.

```python
import os

import boto3
import cfnresponse

client = boto3.client('lambda')
responseData = {'Data': 'OK'}
function_name_suffix = os.getenv('FUNCTION_NAME_SUFFIX')
s3_bucket_suffix = os.getenv('S3_BUCKET_SUFFIX')
s3_key_suffix = os.getenv('S3_KEY_SUFFIX')


def handler(event, context):
    try:
        print('RequestType')
        print(event['RequestType'])
        if event['RequestType'] == 'Update':
            print(os.linesep.join('Environment Variable {} : {}'.format(k, v) for k, v in os.environ))

            function_vars = [var for var in os.environ if var.endswith(function_name_suffix)]

            for function_name in function_vars:
                function_name_prefix = function_name.replace(function_name_suffix, '')
                bucket_var = '{}{}'.format(function_name_prefix, s3_bucket_suffix)
                key_var = '{}{}'.format(function_name_prefix, s3_key_suffix)

                s3_bucket = os.getenv(bucket_var)
                s3_key = os.getenv(key_var)

                print('s3Bucket Key:{} Value:{}'.format(bucket_var, s3_bucket))
                print('s3Key Key:{} Value:{}'.format(key_var, s3_key))

                response = client.update_function_code(
                    FunctionName=function_name,
                    S3Bucket=s3_bucket,
                    S3Key=s3_key,
                    Publish=True
                )
                print('response: {}'.format(response))
        return cfnresponse.send(event, context, cfnresponse.SUCCESS, responseData, 'S3PutObjectFunction')
    except Exception as e:
        print(e)
        return cfnresponse.send(event, context, cfnresponse.FAILED, responseData, 'S3PutObjectFunction')
```

File: wps-cloudformation-template.yaml [hunk 2]
```diff
@@ -753,6 +757,104 @@ Resources:
...
+
+                             response = client.update_function_code(
+                                 FunctionName=function_name,
+                                 S3Bucket=s3_bucket,
+                                 S3Key=s3_key,
+                                 Publish=True
+                             )
+                             print('response: {}'.format(response))
+
+                 return cfnresponse.send(event, context, cfnresponse.SUCCESS, responseData, 'S3PutObjectFunction')
```
Comment (lwgordonimos) [inline wps-cloudformation-template.yaml]: Is this supposed to refer to the current function, and not 'S3PutObjectFunction'?

File: wps-cloudformation-template.yaml (hunk 1 above)
Comment (bhasin85) [reply wps-cloudformation-template.yaml]: Updated and tested the code

File: wps-cloudformation-template.yaml (hunk 2 above)
Comment (bhasin85) [reply wps-cloudformation-template.yaml]: Nop... I have fixed it

**Comment by bhasin85** [comment]
@lwgordonimos  Addressed review comments :+1:

File: wps-cloudformation-template.yaml [hunk 3]
```diff
@@ -753,6 +757,101 @@ Resources:
...
+      ManagedPolicyArns:
+        - 'arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole'
+      Policies:
+        - PolicyName: UpdateVersion
+          PolicyDocument:
+            Version: 2012-10-17
+            Statement:
+              - Effect: Allow
+                Action:
+                  - 'lambda:UpdateFunctionCode'
```
Comment (lwgordonimos) [inline wps-cloudformation-template.yaml]: Can this access be narrowed at all to not be able to update any function?

The biggest concern I have about these custom resources is that they escape from the constraints of the template and are given access to arbitrary resources (as opposed to only the resources included in the template, addressed by reference).

Is there some way to limit this at least by the name of the function, or even better by reference (not sure if this creates a circular dependency however)?

Comment (bhasin85) [reply wps-cloudformation-template.yaml]: @lwgordonimos  Agree... I should have thought about it. 

Made the needed changes :+1:

Comment (lwgordonimos) [reply wps-cloudformation-template.yaml]: Nice! Makes me feel a lot less nervous about the custom resources.

**Pull request merged**